### PR TITLE
Feature/Release ARM Image

### PR DESCRIPTION
**Pull request by iGxnon**
Please briefly answer these questions:

The test workflow: https://github.com/iGxnon/rust-ci-test/actions/runs/5362996116

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

  Etcd images are compatible with computers that have ARM architecture. Additionally, xline can also support ARM thanks to Rust's cross-compilation.

* what changes does this pull request make?

  This pull request updated the release's CI to allow for automatic building of ARM images.
  
  - Added ARM image building process in Github Action CI
  - Added release artifacts for macOS and ARM Linux
  - Reorganize quick-start readme

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

  No

### CODE REVIEW

File: doc/quick-start/README.md [hunk 1]
```diff
@@ -1,52 +1,19 @@
...
-## Benchmark
-
-```bash
-./scripts/benchmark.sh
+docker run -it --name=xline datenlord/xline \
+  xline \
+  --name xline \
+  --storage-engine rocksdb \
+  --members xline=127.0.0.1:2379 \
+  --data-dir /usr/local/xline/data-dir
```
Comment (iGxnon) [inline doc/quick-start/README.md:13]: According to #335, Xline currently does not support single node cluster.
This PR should be merged once the single node cluster implementation is completed.

File: doc/quick-start/README.md [hunk 2]
```diff
@@ -86,13 +53,70 @@ cd Xline
 cargo build --release
 ```
 
+### Run Xline
+
+```bash
+./target/release/xline --name xline \
+  --storage-engine rocksdb \
+  --members xline=127.0.0.1:2379 \
+  --data-dir <path-to-data-dir>
```
Comment (iGxnon) [inline doc/quick-start/README.md:62]: Same as above

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -5,45 +5,134 @@ on:
...
+  REGISTRY_IMAGE: datenlord/xline
 
 jobs:
-  docker:
-    name: Docker
-    runs-on: ubuntu-latest
+  push_image:
+    name: Push Docker Image
+    runs-on: ${{ matrix.job.os }}
+    # continue-on-error: true
```
Comment (themanforfree) [inline .github/workflows/docker.yml:14]: If we didn't need this, please remove

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -38,36 +26,163 @@ jobs:
...
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
+      - name: Install Packages (macOS)
+        if: matrix.job.os == 'macos-latest'
+        run: |
+          .github/scripts/install_macos_deps.sh
+
+      - name: Install Toolchain
+        uses: dtolnay/rust-toolchain@master
```
Comment (themanforfree) [inline .github/workflows/release.yml:84]: Same as above

File: .github/workflows/docker.yml [hunk 5]
```diff
@@ -5,45 +5,134 @@ on:
...
+        with:
+          context: .github/scripts
+          file: .github/scripts/Dockerfile.${{ matrix.job.target }}
+          tags: ${{ matrix.job.cross_image }}:latest
+          load: true
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
+      - name: Install Toolchain
+        uses: dtolnay/rust-toolchain@master
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply .github/workflows/docker.yml:51]: `actions-rs/toolchain@v1` seems to have not been updated for a long time.

File: .github/workflows/docker.yml [hunk 6]
```diff
@@ -5,45 +5,134 @@ on:
...
-          command: build
-          args: --release
-      - run: |
-          mv ./target/release/xline ./scripts
-          mv ./target/release/benchmark ./scripts
+          cache-on-failure: true
+          key: '${{ matrix.job.target }}'
+
+      - name: Install cross
+        uses: baptiste0928/cargo-install@v2
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply .github/workflows/docker.yml:62]: I don't know what exactly happened when using `use-cross: true` in actions-rs/cargo (some logs were not displayed in the Github Action), and it seems to cause cross-compilation failure for rocksdb ?
This version happens to work properly.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (iGxnon) [reply .github/workflows/docker.yml]: done

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -38,36 +26,162 @@ jobs:
...
+      - name: Upload Digest
+        uses: actions/upload-artifact@v3
+        if: startsWith(matrix.job.platform, 'linux/')
+        with:
+          name: digests
+          path: /tmp/digests/*
+          if-no-files-found: error
+          retention-days: 1
+
+  merge_image:
```
Comment (Phoenix500526) [inline .github/workflows/release.yml:157]: This `merge_image` is identical to the one in docker.yml. This is a kind of duplication. Can we extract it into a separate yaml file and reuse it?

File: .github/workflows/docker.yml [hunk 8]
```diff
@@ -5,45 +5,133 @@ on:
...
+
+      - name: Upload Digest
+        uses: actions/upload-artifact@v3
+        with:
+          name: digests
+          path: /tmp/digests/*
+          if-no-files-found: error
+          retention-days: 1
+
+  merge_image:
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply .github/workflows/docker.yml:107]: The `merge_image` job is used to combine images with different digests (which were previously built) into **a single tag**, and then push this tag into DockerHub.
After that, when users pull this tag, they can automatically choose the appropriate digests image.

File: .github/workflows/release.yml (hunk 7 above)
Comment (iGxnon) [reply .github/workflows/release.yml:157]: If we are using something like a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) to separate these steps, we may need to checkout the xline repository once again to get this `<action>.yml` file, while in `merge_image` job we have no reason to checkout xline repository.

File: doc/quick-start/README.md (hunk 1 above)
Comment (iGxnon) [reply doc/quick-start/README.md:13]: This PR should rebase on #346

File: .github/workflows/docker.yml [hunk 9]
```diff
@@ -5,45 +5,133 @@ on:
...
+        with:
+          context: .github/scripts
+          file: .github/scripts/Dockerfile.${{ matrix.job.target }}
+          tags: ${{ matrix.job.cross_image }}:latest
+          load: true
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
+      - name: Install Toolchain
+        uses: dtolnay/rust-toolchain@master
```
Comment (bsbds) [inline .github/workflows/docker.yml:50]: I think we should consider using stable version instead of master

File: .github/workflows/docker.yml [hunk 10]
```diff
@@ -5,45 +5,133 @@ on:
...
   CI_RUST_TOOLCHAIN: 1.70.0
+  REGISTRY_IMAGE: datenlord/xline
 
 jobs:
-  docker:
-    name: Docker
-    runs-on: ubuntu-latest
+  push_image:
+    name: Push Docker Image
+    runs-on: ${{ matrix.job.os }}
```
Comment (bsbds) [inline .github/workflows/docker.yml:13]: While we can run jobs on vm that gihub provides, why do we need the `cross` crate for compilation?

Comment (iGxnon) [reply .github/workflows/docker.yml:13]: Currently, [GitHub-hosted runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners) are all x86_64 architecture. If we need to support ARM, we need to cross compile, or use ARM architecture Self-hosted runners.

File: .github/workflows/docker.yml (hunk 9 above)
Comment (iGxnon) [reply .github/workflows/docker.yml]: done

**Review by Phoenix500526 — APPROVED** [review]

**Review by themanforfree — DISMISSED** [review]

**Review by themanforfree — APPROVED** [review]

**Pull request merged**